profiling/server.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package profiling
+
+import (
+	"net/http"
+	"net/http/pprof"
+	"strconv"
+	"sync"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// profilingPort is the port where we expose profiling information if profiling is enabled
+	profilingPort = ":8008"
+
+	// profilingKey is the name of the key in config-observability config map that indicates whether profiling
+	// is enabled of disabled
+	profilingKey = "profiling.enable"
+)
+
+// Handler holds the main HTTP handler and a flag indicating
+// whether the handler is active
+type Handler struct {
+	enabled    bool
+	enabledMux sync.Mutex
+	handler    http.Handler
+	log        *zap.SugaredLogger
+}
+
+// NewHandler create a new ProfilingHandler which serves runtime profiling data
+// according to the given context path
+func NewHandler(logger *zap.SugaredLogger) *Handler {
+	const pprofPrefix = "/debug/pprof/"
+
+	mux := http.NewServeMux()
+	mux.HandleFunc(pprofPrefix, pprof.Index)
+	mux.HandleFunc(pprofPrefix+"cmdline", pprof.Cmdline)
+	mux.HandleFunc(pprofPrefix+"profile", pprof.Profile)
+	mux.HandleFunc(pprofPrefix+"symbol", pprof.Symbol)
+	mux.HandleFunc(pprofPrefix+"trace", pprof.Trace)
+
+	return &Handler{
+		enabled: false,
+		handler: mux,
+		log:     logger,
+	}
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.enabledMux.Lock()
+	defer h.enabledMux.Unlock()
+	if h.enabled {
+		h.handler.ServeHTTP(w, r)
+	} else {
+		http.NotFoundHandler().ServeHTTP(w, r)
+	}
+}
+
+// UpdateFromConfigMap modifies the Enabled flag in the Handler
+// according to the value in the given ConfigMap
+func (h *Handler) UpdateFromConfigMap(configMap *corev1.ConfigMap) {
+	profiling, ok := configMap.Data[profilingKey]
+	if !ok {
+		return
```

Comment:
Is there a default value we should assume?  e.g. does omission equal enabled or disabled?

Reply:
One though, without a default here the same config could result in two different behaviors across pod restarts.

Reply:
Ah. Good point. Currently, the default value is false at the beginning. Later, when the configmap is updated during runtime it will try to read the value and update it accordingly. If omitted, the current value remains. It's true that when the configmap is updated to profiling.enable=true and the pod is then restarted it will be again false (even if the configmap has profiling.enable=true) until someone updates the configmap again.

I can add initialization so that the pod reads the current value from the configmap at the beginning. And if omitted then set it to false.